spectralmatch/utils/utils_global.py

```diff
@@ -61,34 +61,160 @@ def _calculate_image_stats(
+        bound_j (dict): Bounds of the second image in the format
+            {"x_min", "x_max", "y_min", "y_max"}.
         nodata_i (float): The NoData value for the first image.
         nodata_j (float): The NoData value for the second image.
+        vector_mask_path (str, optional): File path to a GPKG containing mask
+            polygons. Only polygons whose 'image' attribute matches the
+            base name of the image file (excluding extension) are used
+            to define valid pixels. Pixels outside these polygons are
+            set to the image’s NoData value. If None, no masking is applied.
 
     Returns:
         tuple: A tuple containing:
-        - overlap_stat: Dictionary of overlap statistics in the format:
-        {id_i: {id_j: {band: {'mean': value, 'std': value, 'size': value}}},
-        id_j: {id_i: {band: {'mean': value, 'std': value, 'size': value}}}}
-        - whole_stats: Dictionary of whole image statistics in the format:
-        {id_i: {band: {'mean': value, 'std': value, 'size': value}}}
+            - overlap_stat: Dictionary of overlap statistics in the format:
+              {
+                id_i: {
+                  id_j: {
+                    band: {'mean': value, 'std': value, 'size': value}
+                  }
+                },
+                id_j: {
+                  id_i: {
+                    band: {'mean': value, 'std': value, 'size': value}
+                  }
+                }
+              }
+            - whole_stats: Dictionary of whole image statistics in the format:
+              {
+                id_i: {
+                  band: {'mean': value, 'std': value, 'size': value}
+                },
+                id_j: {
+                  band: {'mean': value, 'std': value, 'size': value}
+                }
+              }
     """
     # Initialize the result dictionaries
     overlap_stat = {id_i: {id_j: {}}, id_j: {id_i: {}}}
     whole_stats = {id_i: {}, id_j: {}}
 
+    # --- Helper function to rasterize a relevant mask from GPKG and apply it ---
+    def apply_vector_mask_if_needed(dataset, image_path):
+        """
+        If vector_mask_path is not None, rasterize the polygons from the GPKG where
+        the 'image' field contains the base_name of the given image_path (split by ';'),
+        and return a mask array (1 where valid). Pixels outside those polygons
+        are set to NoData in the caller.
+
+        Args:
+            dataset (gdal.Dataset): The opened raster dataset (for size/projection).
+            image_path (str): Path to the raster, used to derive the base_name.
+
+        Returns:
+            np.ndarray or None: A mask array (same shape as raster) with 1 for valid
+                                pixels and 0 for invalid. If no matching features or
+                                vector_mask_path is None, returns None.
+        """
+        if vector_mask_path is None:
+            return None  # No masking needed
+
+        import os
```

Comment:
in principal, I think that all imports should be at the top of the file and not in functions. This really for readability and following PEP standards. The only time to import within functions will be to avoid circular imports. But, you dont have any here. So, best to remove these imports here and add any that are missing to the top of the file.